### ISSUE TITLE
Remove the 255 cap from the Health HUD

### ISSUE THREAD

**Issue by basuritashka** [ValveSoftware/halflife#2149]
Title. If you have more than > 255 HP your HUD wont display more than that value.

**Comment by SamVanheer** [comment]
This is because the network message `Health` (global `gmsgHealth`) sends the value as a byte. Sending it as a short should allow for much larger health values, but would require changes to be made to any third party plugins that intercept the message (i.e. AMXMod).

**Comment by shelru** [comment]
I think' there must be unsigned int.

**Comment by eat1k** [comment]
It would be very nice.

**Comment by wopox1337** [comment]
@shelru mb `unsigned short int`?

**Comment by 2010kohtep** [comment]
It can also break backward compatibility with old clients that are used in no-steam versions of the game, which can cause an outflow of players, since there are no proper ways to change ```Health```. It may be worth creating a new message ```Health2```, in which the int32 value of health will be transmitted. This will not create problems.

**Comment by d3m37r4** [comment]
Split radar from the HUD HP and AP would not be superfluous to be able to hide the radar separately (via amxx plugin).

**Comment by Uiharu** [comment]
To be fair no steam versions shouldn't be supported, to begin with.

However, I still agree that the safest bet would be creating a new "health2"

**Comment by SamVanheer** [comment]
> 
> 
> It can also break backward compatibility with old clients that are used in no-steam versions of the game, which can cause an outflow of players, since there are no proper ways to change `Health`. It may be worth creating a new message `Health2`, in which the int32 value of health will be transmitted. This will not create problems.

This message exists in game code, so if you're using a server dll that has this you should also be using a client dll that matches it. Plus, as @Uiharu said, no-Steam is not supported by Valve and should not be considered in any decisions.

If open sourcing is possible then no-Steam as it exists now will no longer matter anyway.

**Comment by shelru** [comment]
+1, non-steam shouldn't be supp.

**Comment by 2010kohtep** [comment]
@SamVanhee,

The presence of non-existent user messages on the client is optional. The callback list is initialized in the client library, which calls ```HookServerMsg``` function, which sets a callback for each message. If message is not set by client library, the message will remain uninitialized, but it will not affect the client. Due to this, the work of the current protocol is not violated if we use the ```Health2``` solution.

Although no-steam should not be supported by official GoldSrc developers (with which I, in general, agree), but server online in the gaming community also has value, since most players prefer no-steam instead of steam. Changing the protocol can cause some excitement in the community. That's why I think this should also be taken into account when working on updates.

**Comment by shelru** [comment]
У него никнейм с r на конце.

**Comment by SamVanheer** [comment]
> 
> 
> @SamVanheer,
> 
> The presence of non-existent user messages on the client is optional. The callback list is initialized in the client library, which calls `HookServerMsg` function, which sets a callback for each message. If message is not set by client library, the message will remain uninitialized, but it will not affect the client. Due to this, the work of the current protocol is not violated if we use the `Health2` solution.
> 
> Although no-steam should not be supported by official GoldSrc developers (with which I, in general, agree), but server online in the gaming community also has value, since most players prefer no-steam instead of steam. Changing the protocol can cause some excitement in the community. That's why I think this should also be taken into account when working on updates.

I agree with the Health2 solution for Half-Life itself on account of mods that only have a custom server dll requiring backwards compatibility, but other games don't need it.

There is only one Counter-Strike, TFC, etc so if you're running the wrong client dll for one of those then you need to update, regardless of whether you're running a Steam or no-Steam version.
No-Steam, WON2, Xash, etc are all in violation of the EULA so Valve need not provide any support for those.

I hope that the engine and the games will be open sourced, providing a legal path to making versions of these games that do not require Steam auth. That should be the only way to play them that is in any way "No-Steam".

**Comment by shelru** [comment]
@wopox1337 они чего, серьезно вынесли деньги в таб?

**Comment by 2010kohtep** [comment]
@shelru,
Не знаю, не смотрел. mikela сказал, что проблема текстур временная и будет исправлена: #2148

**Comment by WPMGPRoSToTeMa** [comment]
The client already knows the actual value of health from `health` field of `clientdata_t`. I think we can just use it. By default it is encoded with 10 bits (values from -512 to 511 IIRC), but it can be changed in `delta.lst`.

**Comment by SamVanheer** [comment]
True, but networked fields may be changed to the wrong value if the client simulates an old frame, so the HUD may change values for single frames. This should be tested before being added.

**Comment by basuritashka** [comment]
@kisak-valve can you add a Feature Request to this please?

**Comment by mikela-valve** [comment]
I think the safest way to handle this is just to add a health2 message as described and deprecate the health message, to ensure the HUD always has an up to date value for such an important field and to make the least amount of breaking changes possible. 

At least where it doesn’t create a significant amount of otherwise unnecessary work, given the long history and community support for all of the Half-Life games, I’d always rather err on the side of not breaking things that our players may rely on regardless of what they are. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by shelru** [comment]
Alfred was fired from GoldSrc dev. team?

**Comment by basuritashka** [comment]
Does this fix covers the Armor hud too?

**Comment by 2010kohtep** [comment]
Armor value sets via the ```Battery``` user message, and its maximum value is 65535. It does not need to be fixed.

**Comment by wilianmaique** [comment]
CS GO is free, why CS 1.6 not?

**Comment by BlackShadow** [comment]
@mikela-valve We gonna receive this on Half-Life and others? Pretty much all Goldsource game has this.

**Comment by BlackShadow** [comment]
@mikela-valve We gonna receive this on Half-Life and others? Pretty much all Goldsource games has this.

**Comment by RauliTop** [comment]
> Split radar from the HUD HP and AP would not be superfluous to be able to hide the radar separately (via amxx plugin).

You can create anothe issue. I haven't seen any related to this yet.

**Comment by SmilexGamer** [comment]
With this cap removal, it should be also removed the display cap of:
- armor (currently 999);
- round time (currently 9);
- money (currently 999999);
- ammo/clip (currently 255).
Basically, all countable things that are displayed in the HUD.

**Comment by Solokiller** [comment]
Note that increasing the maximum health allowed also requires changes to be made to the hud's drawing code.

The best way to do this is to use a new number drawing function that draws in reverse like this:
```cpp
int CHud::DrawHudNumberReverse(int x, int y, int number, int flags, int r, int g, int b)
{
	if (number > 0 || (flags & DHN_DRAWZERO))
	{
		const int digitWidth = GetSpriteRect(m_HUD_number_0).right - GetSpriteRect(m_HUD_number_0).left;

		int remainder = number;

		do
		{
			const int digit = remainder % 10;
			const int digitSpriteIndex = m_HUD_number_0 + digit;

			//This has to happen *before* drawing because we're drawing in reverse
			x -= digitWidth;

			SPR_Set(GetSprite(digitSpriteIndex), r, g, b);
			SPR_DrawAdditive(0, x, y, &GetSpriteRect(digitSpriteIndex));

			remainder /= 10;
		}
		while (remainder > 0);
	}

	return x;
}
```

Additionally a function to calculate the width in pixels of a number:
```cpp
int CHud::GetHudNumberWidth(int number, int width, int flags)
{
	const int digitWidth = GetSpriteRect(m_HUD_number_0).right - GetSpriteRect(m_HUD_number_0).left;

	int totalDigits = 0;

	if (number > 0)
	{
		totalDigits = static_cast<int>(log10(number)) + 1;
	}
	else if (flags & DHN_DRAWZERO)
	{
		totalDigits = 1;
	}

	totalDigits = V_max(totalDigits, width);

	return totalDigits * digitWidth;
}
```

Then modify this:
https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/cl_dll/health.cpp#L221

To use the new function:
```cpp
//Reserve space for 5 digits
x += gHUD.GetHudNumberWidth(m_iHealth, 5, DHN_DRAWZERO);

gHUD.DrawHudNumberReverse(x, y, m_iHealth, DHN_DRAWZERO, r, g, b);

//x = gHUD.DrawHudNumber(x, y, DHN_3DIGITS | DHN_DRAWZERO, m_iHealth, r, g, b);
```

Additionally modifying hud battery drawing:
https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/cl_dll/battery.cpp#L137

To put the start of it further to the right:
```cpp
x = ScreenWidth/4;
```

On the lowest supported resolution (640x480) this puts the battery icon and value in the same position relative to the health value and vertical bar. On higher resolutions it goes further and further right, but there is plenty of space between it and the ammo hud.

This is how it looks with a value that has 5 digits (2 byte values can have that many):
![image](https://i.imgur.com/kraOuPf.png)

Ideally the same could be done to all hud numeric drawing.

This should be checked with each game to verify that there is no overlap between hud elements specific to them.
